Add scaffold for Modules local package

mokagio opened the pull request:
## Description
This local package will contain all the current local packages and the frameworks at present in dedicated project files. See https://linear.app/a8c/issue/AINFRA-465

## Steps to reproduce
N.A.

## Testing information
See the unit test added to the WooCommerce setup.

## Screenshots


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. — N.A.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

mokagio committed:
Add scaffold for Modules local package

This local package will contain all the current local packages and the
frameworks at present in dedicated project files.

mokagio committed:
Run dummy module tests as part of WooCommerce tests

dangermattic commented:
<!--
  0 Errors
  1 Warning: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio commented:
@woocommerce/ios-developers I'd appreciate if you could have a look at the tests here. They fail consistently on `AppSettingsStoreTests.test_it_can_save_the_installationDate_when_the_settings_file_does_not_exist()` in CI but not on my machine. I don't understand why, but maybe something obvious will jump to your experienced eyes.

Noticed I included #15608 in the changes here, just to make sure that the CI issues was not due to old toolchain.

Thanks!

staskus committed:
Use temporary directory rather than real documents directory in tests

If system disables write access to document directory the test could fail. Use temporary directory instead.

staskus commented:
@mokagio I attempted a quick fix. The CI errors suggested the system failed to write the file in the documents directory, so I used a temporary directory in tests instead. The tests are still running, but it looks like this particular one succeeded. I'll check back on Monday in case there are still failures.

mokagio commented:
Thanks @staskus !

The tests run green. Awesome! ✅ 

I pushed a revert for the tooling upgrade just to make sure the two things are unrelated. If all runs green, I'll cleanup with a rebase then ask for an official review.

mokagio commented:
@woocommerce/ios-developers @woocommerce/apps-infrastructure everything is ready for a review here. it would be much appreciated 🙇‍♂️

AliSoftware commented on .gitignore:
```diff
@@ -99,3 +99,8 @@ WooCommerce/WooCommerceScreenshots/Mocks/vendor/
+
+# Local modules
+Modules/.build
+Moudles/.swiftpm/configuration/registries.json
+Moudles/.swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata
```
```suggestion
Modules/.swiftpm/configuration/registries.json
Modules/.swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata
```

mokagio replied:
🤦‍♂️ 

Thanks @AliSoftware

AliSoftware commented on Storage/Storage/Tools/GeneralAppSettingsStorage.swift:
```diff
@@ -17,8 +18,13 @@ public struct GeneralAppSettingsStorage {
-    public init(fileStorage: FileStorage = PListFileStorage()) {
+    public init(fileStorage: FileStorage = PListFileStorage(), fileURL: URL? = nil) {
         self.fileStorage = fileStorage
+        if let fileURL = fileURL {
+            self.fileURL = fileURL
+        } else {
+            self.fileURL = Constants.generalAppSettingsFileURL
+        }
```
Nitpick unrelated to the Modules scaffolding that's the focus on this PR, but why not just write this instead?
```suggestion
        self.fileURL = fileURL ?? Constants.generalAppSettingsFileURL
```

AliSoftware reviewed (changes requested):
Typo in `.gitignore` which might accidentally let some files through

Otherwise LGTM

mokagio committed:
Fix typo in `.gitignore` `s/Moud/Modu/`

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio committed:
Condense an if-else into an nil coalescing

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

AliSoftware commented:
@mokagio you might want to also adjust the `Dangerfile` to fix the paths of the `Package.swift` and `Package.resolved` that it checks (using your discoveries from paNNhX-Yc-p2) to avoid it [this issue](https://github.com/woocommerce/woocommerce-ios/pull/15597#issuecomment-2858069641). Though I guess you can do that in a follow-up PR (after creating a dedicated Linear issue under [your project](https://linear.app/a8c/project/adopt-centralize-modules-swift-package-setup-in-woocommerce-ios-fd0e002a2b6e/issues))

<img width="853" alt="image" src="https://github.com/user-attachments/assets/fdb44330-bc63-49dc-b530-28c49c3e4eef" />
